lisp/lisplib/libhelp/libhelp.go

```diff
@@ -0,0 +1,199 @@
+		env.AddSpecialOps(true, op)
+	}
+	return lisp.Nil()
+}
+
+var ops = []*libutil.Builtin{
+	libutil.FunctionDoc("help", lisp.Formals("var-name"), opHelp,
+		`
+		Prints documentation for the given variable name.  Functions have their
+		signature and any docstring rendered.  Other variables have their types
+		and current values printed.
+		`),
+	libutil.FunctionDoc("help-package", lisp.Formals("pkg-name"), opHelpPackage,
+		`
+		Prints documentation for exported symbols in the specified package.
+		`),
+	libutil.FunctionDoc("help-package-symbols", lisp.Formals("pkg-name", lisp.OptArgSymbol, "all"), opPackageSymbols,
+		`
+		Prints symbols defined in the specified package.  If a second argument
+		is given which evaluates as true then unexported symbols in the package
+		will also be printed.
+		`),
+}
+
+func opHelp(env *lisp.LEnv, args *lisp.LVal) *lisp.LVal {
+	name := args.Cells[0]
+	if name.Type != lisp.LSymbol {
+		return env.Errorf("argument is not a symbol: %v", lisp.GetType(name))
+	}
+	err := RenderVar(env.Runtime.Stderr, env, name.Str)
+	if err != nil {
+		return env.Error(err)
+	}
+	return lisp.Nil()
+}
+
+func opHelpPackage(env *lisp.LEnv, args *lisp.LVal) *lisp.LVal {
+	name := args.Cells[0]
+	if name.Type != lisp.LSymbol {
+		return env.Errorf("argument is not a symbol: %v", lisp.GetType(name))
+	}
+	err := RenderPkgExported(env.Runtime.Stderr, env, name.Str)
+	if err != nil {
+		return env.Error(err)
+	}
+	return lisp.Nil()
+}
+
+func opPackageSymbols(env *lisp.LEnv, args *lisp.LVal) *lisp.LVal {
+	name := args.Cells[0]
+	if name.Type != lisp.LSymbol {
+		return env.Errorf("argument is not a symbol: %v", lisp.GetType(name))
+	}
+	printAll := env.Eval(args.Cells[1])
+	if printAll.Type == lisp.LError {
+		return printAll
+	}
+	pkg := env.Runtime.Registry.Packages[name.Str]
+	if pkg == nil {
+		return env.Errorf("no package: %q", name)
+	}
+	if lisp.True(printAll) {
+		for sym := range pkg.Symbols {
+			_, err := fmt.Fprintln(env.Runtime.Stderr, sym)
+			if err != nil {
+				return env.Error(err)
+			}
+		}
+	} else {
+		exports := pkg.Externals
+		for _, exsym := range exports {
+			_, err := fmt.Fprintln(env.Runtime.Stderr, exsym)
+			if err != nil {
+				return env.Error(err)
+			}
+		}
+	}
+	return lisp.Nil()
+}
+
+func FunDocstring(v *lisp.LVal) string {
+	if v.Builtin() != nil {
+		if len(v.Cells) > 1 {
+			return v.Cells[1].Str
+		}
+		return ""
+	}
+	// Functions of the form (lambda (x) "abc") are considered constant string
+	// functions without documentation so there must be a length check on the
+	// function body.
+	if len(v.Cells) > 2 && v.Cells[1].Type == lisp.LString {
+		return v.Cells[1].Str
+	}
+	return ""
+}
+
+func RenderPkgExported(w io.Writer, env *lisp.LEnv, query string) error {
+	pkg := env.Runtime.Registry.Packages[query]
+	if pkg == nil {
+		return fmt.Errorf("no package: %q", query)
+	}
+	exports := pkg.Externals
+	for i, exsym := range exports {
+		if i > 0 {
+			_, err := fmt.Fprintln(w)
```

Comment:
Looks like this produces extra newlines, especially when a function has no documentation.

Reply:
Having looked at the output again I think it's better looking this way and I think a little more in line with python's `help` function. I guess one thing python does is group variables (non-functions).

If you remove the extra newlines I don't really think it looks good or consistent when you have a mix of document and undocumented procedures. Maybe you are imagining something a little different from me.

The way I see it this text rendering isn't really meant to be held to any backcompat requirements for the time being (though I am fixing the nondeterminism you mentioned). I think we should be free to change the rendering/ordering and do something a bit fancier in the future. I'll just document that for now.